src/main/java/xfacthd/framedblocks/common/compat/jei/camo/CamoCraftingHelper.java

```diff
@@ -0,0 +1,158 @@
+
+    public CamoCraftingHelper()
+    {
+        this.helperRecipe = new CamoApplicationRecipe(CraftingBookCategory.MISC, Ingredient.of(Items.BRUSH));
+        this.camoExamplesIngredient = Ingredient.of(JEI_CAMO_BLOCK_EXAMPLES);
+    }
+
+    public void scanForItems(IIngredientManager ingredientManager)
+    {
+        List<ItemStack> camoExamples = new ArrayList<>();
+        List<ItemStack> framedBlocks = new ArrayList<>();
+
+        for (ItemStack stack : ingredientManager.getAllItemStacks())
+        {
+            if (camoExamples.size() < MAX_CAMO_EXAMPLE_INGREDIENTS_COUNT)
+            {
+                CamoContainerFactory<?> factory = getCamoContainerFactory(stack);
+                if (factory != null)
+                {
+                    camoExamples.add(stack);
+                }
+            }
+
+            IFramedBlock framedBlock = getFramedBlock(stack);
+            if (framedBlock != null)
+            {
+                framedBlocks.add(stack);
+            }
+        }
+
+        this.camoExamples = Collections.unmodifiableList(camoExamples);
+        this.framedBlocks = Collections.unmodifiableList(framedBlocks);
+    }
+
+    public List<ItemStack> getFramedBlocks()
+    {
+        return framedBlocks;
+    }
+
+    @Nullable
+    public static CamoContainerFactory<?> getCamoContainerFactory(ItemStack itemStack)
+    {
+        CamoContainerFactory<?> factory = CamoContainerHelper.findCamoFactory(itemStack);
+        if (factory == null || !factory.canApplyInCraftingRecipe(itemStack))
+        {
+            return null;
+        }
+        return factory;
+    }
+
+    @Nullable
+    public IFramedBlock getFramedBlock(ItemStack itemStack)
+    {
+        if (itemStack.getItem() instanceof BlockItem item && item.getBlock() instanceof IFramedBlock framedBlock)
+        {
+            return framedBlock;
+        }
+        return null;
+    }
+
+    public boolean isDoubleFramedBlock(ItemStack itemStack)
+    {
+        IFramedBlock framedBlock = getFramedBlock(itemStack);
+        return framedBlock != null && isDoubleFramedBlock(framedBlock);
+    }
+
+    public boolean isDoubleFramedBlock(IFramedBlock framedBlock)
+    {
+        return framedBlock.getBlockType().consumesTwoCamosInCamoApplicationRecipe();
+    }
+
+    public ItemStack calculateOutput(ItemStack frame, ItemStack inputOne, ItemStack inputTwo)
+    {
+        Minecraft minecraft = Minecraft.getInstance();
+        ClientLevel level = minecraft.level;
+        assert level != null;
+        RegistryAccess registryAccess = level.registryAccess();
+
+        List<ItemStack> inputs = List.of(frame, new ItemStack(Items.BRUSH), inputOne, inputTwo);
+        CraftingInput craftingInput = CraftingInput.of(2, 2, inputs);
+        return helperRecipe.assemble(craftingInput, registryAccess);
+    }
+
+    public List<ItemStack> getCamoExampleStacks(Ingredient ingredient)
+    {
+        if (ingredient.equals(camoExamplesIngredient))
```

Comment:
I tried doing reference equality checking here but it breaks when a bookmarked recipe is read from a file.
The new recipe created by the codec doesn't have the exact same ingredient.

The deep equality check will compare the ingredient's `value` field, which is a `TagValue`, which uses the tag's resource location. Since it's not just comparing items, this thankfuly means it will not match undesired ingredients.